### PR TITLE
Refactor vring abstraction

### PR DESCRIPTION
**Pull request by jiangliu**
Add `trait VringT` and provide three implementation of it: VringState, VringMutex, VringRwLock,
so clients could choose different implementation.

### CODE REVIEW

File: src/vring.rs [hunk 1]
```diff
@@ -7,16 +7,133 @@
...
 use std::os::unix::io::{FromRawFd, IntoRawFd};
 use std::result::Result;
-use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::sync::{Arc, Mutex, MutexGuard, RwLock, RwLockReadGuard, RwLockWriteGuard};
 
 use virtio_queue::{Error as VirtQueError, Queue};
 use vm_memory::{GuestAddress, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
 use vmm_sys_util::eventfd::EventFd;
 
+/// Struct to hold an shared reference to the underlying `VringState` object.
```
Comment (stefano-garzarella) [inline src/vring.rs:19]: s/an/a

File: src/vring.rs [hunk 2]
```diff
@@ -7,16 +7,133 @@
...
+}
+
+pub trait VringT<M: GuestAddressSpace> {
+    /// Create a new instance of Vring.
+    fn new(mem: M, max_queue_size: u16) -> Self;
+
+    /// Get an immutable reference to the kick event fd.
+    fn get_ref(&self) -> VringStateGuard<M>;
+
+    /// Get an immutable reference to the kick event fd.
```
Comment (stefano-garzarella) [inline src/vring.rs:80]: s/immutable/mutable

File: src/vring.rs [hunk 3]
```diff
@@ -7,16 +7,133 @@
...
+    /// Enable/disable queue event index feature.
+    fn set_queue_event_idx(&mut self, enabled: bool);
+
+    /// Set queue enabled state.
+    fn set_queue_ready(&mut self, ready: bool);
+
+    /// Set `EventFd` for kick.
+    fn set_kick(&mut self, file: Option<File>);
+
+    /// Read event from the kick EventFd.
```
Comment (stefano-garzarella) [inline src/vring.rs:122]: Maybe add `` to EventFd as other comments?

**Review by stefano-garzarella — CHANGES_REQUESTED** [review]
The refactor looks great, just a couple of simple copy&paste typos :-)

**Review by stefano-garzarella — APPROVED** [review]

**Comment by slp** [comment]
Thanks for working on this. I must be missing something, but I can't find a way to access the underlying `Queue` in `VringState` to iterate over it. What's the pattern to lock the `Vring` and iterate on the `Queue` to get the descriptor chains?

**Comment by jiangliu** [comment]
> Thanks for working on this. I must be missing something, but I can't find a way to access the underlying `Queue` in `VringState` to iterate over it. What's the pattern to lock the `Vring` and iterate on the `Queue` to get the descriptor chains?

I realized this issue and have updated the patch yesterday. Now we have "VringState::get_queue()" and "VringState::get_queue_mut()" for that purpose.
https://github.com/rust-vmm/vhost-user-backend/pull/30/files#diff-04f224919b2745af8bfb2d2aea4be7d9100e861a8a16f2cd0263cb1021b44c4dR151

**Comment by jiangliu** [comment]
> > > Thanks for working on this. I must be missing something, but I can't find a way to access the underlying `Queue` in `VringState` to iterate over it. What's the pattern to lock the `Vring` and iterate on the `Queue` to get the descriptor chains?
> > 
> > 
> > I realized this issue and have updated the patch yesterday. Now we have "VringState::get_queue()" and "VringState::get_queue_mut()" for that purpose.
> > https://github.com/rust-vmm/vhost-user-backend/pull/30/files#diff-04f224919b2745af8bfb2d2aea4be7d9100e861a8a16f2cd0263cb1021b44c4dR151
> 
> Great, thanks!
> 
> I've tested this crate with virtiofsd, and was able to implement both a multi-threaded method and a single-threaded method for processing the queue, so I think this model should work for most consumers of this crate.
> 
> One change that I think could be interesting, is publishing `VringStateGuard` and `VringStateMutGuard` in the root of the crate (since `vring` is private). That allows consumers to acquire a reference to `VringState` (and thus, the underlying lock, if using `VringMutex` or `VringRwLock`) once and pass it around to other methods. What do you think about this?

Done:)

**Comment by slp** [comment]
Thanks again! :-)

**Review by slp — DISMISSED** [review]

**Comment by slp** [comment]
@jiangliu Could you please rebase this on top of f0af5f7 ?

**Review by slp — APPROVED** [review]
LGTM (again), thanks!

**Review by stefano-garzarella — APPROVED** [review]

**Pull request merged**